pkg/client/client.go

```diff
@@ -618,85 +615,43 @@ func NewClient(ctrlRuntimeClient client.Client, secretName, namespace, regionID
-	if len(secret.Data[kubeRoleArn]) > 0 && len(secret.Data[kubeRoleSessionName]) > 0 && len(secret.Data[kubeRoleSessionExpiration]) > 0 {
-		if roleSessionExpiration, err := utils.String2IntPointer(string(secret.Data[kubeRoleSessionExpiration])); roleSessionExpiration != nil && err == nil {
-			config.RoleArn = string(secret.Data[kubeRoleArn])
-			config.RoleSessionName = string(secret.Data[kubeRoleSessionName])
-			config.RoleSessionExpiration = roleSessionExpiration
+	var secret corev1.Secret
+	if err := ctrlRuntimeClient.Get(context.Background(), client.ObjectKey{Namespace: namespace, Name: secretName}, &secret); err != nil {
+		if apimachineryerrors.IsNotFound(err) {
+			return nil, machineapiapierrors.InvalidMachineConfiguration("alibabacloud credentials secret %s/%s: %v not found", namespace, secretName, err)
 		}
+		return nil, err
 	}
+	return fetchCredentialsIniFromSecret(&secret)
+}
 
-	// roleName
-	if len(secret.Data[kubeRoleName]) > 0 {
-		config.RoleName = string(secret.Data[kubeRoleName])
-	}
-
-	return nil
-}
-
-// fetchCredentialsFileFromConfigMap fetch oleArn, roleName ,roleSessionName ,roleSessionExpiration from configmap
-// roleArn, roleName ,roleSessionName ,roleSessionExpiration
-func fetchCredentialsFileFromConfigMap(namespace string, configManagedClient client.Client, config *providers.Configuration) error {
-	cm := &corev1.ConfigMap{}
-	switch err := configManagedClient.Get(
-		context.Background(),
-		client.ObjectKey{Namespace: namespace, Name: kubeCloudConfigName},
-		cm,
-	); {
-	case apimachineryerrors.IsNotFound(err):
-		// no cloud config ConfigMap
-		return nil
-	case err != nil:
-		return fmt.Errorf("failed to get kube-cloud-config ConfigMap: %w", err)
+func fetchCredentialsIniFromSecret(secret *corev1.Secret) (auth.Credential, error) {
+	creds, ok := secret.Data[kubeCredentialCredentials]
+	if !ok {
+		return nil, fmt.Errorf("failed to fetch key 'credentials' in secret data")
 	}
-
-	if len(cm.Data[kubeRoleArn]) > 0 && len(cm.Data[kubeRoleSessionName]) > 0 && len(cm.Data[kubeRoleSessionExpiration]) > 0 {
-		if roleSessionExpiration, err := utils.String2IntPointer(cm.Data[kubeRoleSessionExpiration]); roleSessionExpiration != nil && err == nil {
-			config.RoleArn = cm.Data[kubeRoleArn]
-			config.RoleSessionName = cm.Data[kubeRoleSessionName]
-			config.RoleSessionExpiration = roleSessionExpiration
-		}
+	f, err := ioutil.TempFile("", "alibaba-creds-*")
+	if err != nil {
+		return nil, err
 	}
-
-	if len(cm.Data[kubeRoleName]) > 0 {
-		config.RoleName = cm.Data[kubeRoleName]
+	_, err = f.Write(creds)
+	if err != nil {
+		return nil, err
 	}
-
-	return nil
+	// This lock is used to prevent the environment variable from being updated while we
+	// are using the environment variable to call the Alibaba credential provider chain.
+	mutex.Lock()
+	os.Setenv(provider.ENVCredentialFile, f.Name())
+	defer mutex.Unlock()
+	defer os.Unsetenv(provider.ENVCredentialFile)
+	defer os.Remove(f.Name())
+	defer f.Close()
```

Comment:
These should be under where f is created, otherwise one of the errors in between might result in a file handle being left open